### ISSUE TITLE
[Question] Use multiprocessing Pool in EvalCallback

### ISSUE THREAD

**Issue by dylanprins** [DLR-RM/stable-baselines3#1232]
### ❓ Question

Hi!

During training we evaluate each x amount of steps using our own evaluate function. 
This process is implemented using an EvalCallback. The strange thing is that the callbacks are added to the model during initialization. Meaning that if the model is on the GPU, the callback is also on the GPU (as it seems). Now, we want to use multiprocessing inside our EvalCallback, but since this callback is on the GPU, these processes will be started on the GPU (with way less RAM ;) ) instead of the CPU.

I was wondering whether it would be able to not include the callback directly in the model.learn or keep the callback on the CPU so this whole evaluation process can be executed on the CPU instead of on the GPU.

To clarify, if i create a mp pool in the init of our EvalCallback, all these processes are opened on the GPU. If i start the pool in the _on_step function itself CUDA starts to panic explaining that mp does not work with CUDA. This makes it seem like the process with the callback inside is on the GPU instead of the CPU.

Hope you understand what i'm aiming for

Thank you!! :)

### Checklist

- [X] I have checked that there is no similar [issue](https://github.com/DLR-RM/stable-baselines3/issues) in the repo
- [X] I have read the [documentation](https://stable-baselines3.readthedocs.io/en/master/)
- [X] If code there is, it is minimal and working
- [X] If code there is, it is formatted using the [markdown code blocks](https://help.github.com/en/articles/creating-and-highlighting-code-blocks) for both code and stack traces.

**Comment by araffin** [comment]
Hello,
could you provide a minimal working example to reproduce the issue?

**Comment by dylanprins** [comment]
Hi! first of all, thanks for the quick reply 
Okay yes i made you an dummy 

```
import gym
import multiprocess.pool
from stable_baselines3.common.callbacks import EventCallback
from stable_baselines3 import PPO


class EvalCallback(EventCallback):
    def __init__(
            self,
            verbose: int = 1,
    ):
        super(EvalCallback, self).__init__(None, verbose=verbose)

    def _on_step(self) -> bool:
        def square(number):
            return number ** 2

        with multiprocess.pool.Pool(4) as p:
            for squared in p.map(square, list(range(10))):
                print(squared)


class DummyEnv(gym.Env):
    def __init__(self):
        self.action_space = gym.spaces.Discrete(2)
        self.observation_space = gym.spaces.Box(low=0, high=1, shape=(1,))
        print("Initialized :)")

    def step(self, action):
        print("I'm going to square multiprocessed using my callback :)")
        return [1], 1, False, {}

    def reset(self):
        print("Nothing to reset here")
        return [1]


if __name__ == "__main__":
    dummy_callback = EvalCallback()
    env = DummyEnv()
    model = PPO('MlpPolicy', env, device='auto')
    model.learn(total_timesteps=10, callback=dummy_callback)
```

The thing is, this works when working only with the cpu.
If we run the model on the GPU the callback object is also moved to the GPU. Meaning if we then open a multiprocessing pool there, CUDA starts warning and giving errors. I think the callback is moved to the gpu because we give it to `model.learn()` and the model is already on the GPU of course. We could slightly alter the code to make it work on the gpu without errors:

```
import gym
import multiprocess.pool
from stable_baselines3.common.callbacks import EventCallback
from stable_baselines3 import PPO


class EvalCallback(EventCallback):
    def __init__(
            self,
            verbose: int = 1,
    ):
        super(EvalCallback, self).__init__(None, verbose=verbose)
        self.pool = miltiprocess.pool.Pool(4)

    def _on_step(self) -> bool:
        def square(number):
            return number ** 2

            for squared in self.pool.map(square, list(range(10))):
                print(squared)


class DummyEnv(gym.Env):
    def __init__(self):
        self.action_space = gym.spaces.Discrete(2)
        self.observation_space = gym.spaces.Box(low=0, high=1, shape=(1,))
        print("Initialized :)")

    def step(self, action):
        print("I'm going to square multiprocessed using my callback :)")
        return [1], 1, False, {}

    def reset(self):
        print("Nothing to reset here")
        return [1]

if __name__ == "__main__":
    dummy_callback = EvalCallback()
    env = DummyEnv()
    model = PPO('MlpPolicy', env, device='auto')
    model.learn(total_timesteps=10, callback=dummy_callback)
```

But the problem with this is that now the multiprocessing pool is opened on the GPU itself. Compared to our normal ram we have way less VRAM (of course).

Thus what i want, is to be able to do a _on_step on the cpu. Meaning the callback should not be fed to an object that is on the GPU. To give an example why and how we use this. We evaluate every 100_000 steps lots of envs using our EvalCallback combined with our own evaluate function. We need custom evaluation because we have specific custom metrics to track. During this evaluation i temporarily take the self.model.policy in the callback and move it to the cpu, perform evaluation on the cpu in a multiprocessing pool and then move it back to the gpu so the model can continue learning there. But the problem here is that the callback object itself is on the GPU, meaning if we open a multiprocessing pool CUDA starts spitting errors.

I hope you understand what i mean, if you have any more questions i'm glad to hear them.

Kind Regards,
Dylan Prins

**Comment by araffin** [comment]
I'm still not sure why the callback would be moved to GPU (it is true that Python will try to do a copy of the runtime env for each process).
From [pytorch doc](https://pytorch.org/docs/stable/notes/multiprocessing.html#cuda-in-multiprocessing):
"The CUDA runtime does not support the fork start method; either the spawn or forkserver start method are required to use CUDA in subprocesses."
In SB3, when using multiple envs and `SubprocVecEnv`, we use `spawn` and `forkserver` as default, and I didn't have any issues so far (using multiple envs for perodic evaluation in the RL Zoo with `SubprocVecEnv`).

The issue here is that because of the pool, you are copying the model to each process, you should probably only send what need to parallelized to the processes and use batch prediction in the main process (what we currently do when evaluating with multiple envs).

> We could slightly alter the code to make it work on the gpu without errors:

this looks better because you don't have the overhead of launching processes everytime.

**Comment by dylanprins** [comment]
Hi! Thanks for the reply.

I have found the issue. I did copy only the parts that I needed for evaluation. So I did a `deepcopy(self.model.policy).cpu()` which was the only thing sent to the pool which could be on the GPU

Still, a `core_count` of processes were started on the GPU. They were not used at all so I think they were just empty CUDA context containing processes.

I checked every tensor in that policy copy, they were all on the cpu. So how could it be that python still thinks it needs to start processes on the GPU? I have no idea, but the fix was simply creating a new temporary policy, and loading in a copy of the current weights. If you only create the policy itself, this should not take a long time.
